File: reverse/ghidra_scripts/At3Dump.py
# Ghidra Jython script: dump analysis artifacts to JSON for AI ingestion
# Usage (headless): -postScript At3Dump.py <output_dir>

import os
import json

from ghidra.program.model.data import StringDataInstance
from ghidra.program.model.symbol import SymbolType
from ghidra.app.decompiler import DecompInterface


def write_json(out_dir, filename, obj):
    path = os.path.join(out_dir, filename)
    f = open(path, "w")
    try:
        json.dump(obj, f, indent=2)
    finally:
        f.close()


def dump_overview(program, out_dir):
    info = {
        "name": program.getName(),
        "language": str(program.getLanguageID()),
        "compiler": str(program.getCompilerSpec().getCompilerSpecID()),
        "image_base": str(program.getImageBase()),
        "entry_point": str(program.getEntryPoint()),
    }
    write_json(out_dir, "overview.json", info)


def dump_strings(program, out_dir):
    listing = program.getListing()
    strings = []
    max_strings = 5000
    it = listing.getDefinedData(True)
    while it.hasNext() and len(strings) < max_strings:
        data = it.next()
        try:
            sdi = StringDataInstance(data)
            value = sdi.getStringValue()
            if value is None:
                continue
            strings.append({
                "address": str(data.getAddress()),
                "value": value,
                "length": sdi.getStringLength(),
            })
        except Exception:
            continue
    write_json(out_dir, "strings.json", strings)


def dump_functions(program, out_dir):
    fm = program.getFunctionManager()
    funcs = []
    it = fm.getFunctions(True)
    while it.hasNext():
        f = it.next()
        body = f.getBody()
        funcs.append({
            "name": f.getName(),
            "entry": str(f.getEntryPoint()),
            "signature": str(f.getSignature()),
            "param_count": f.getParameterCount(),
            "return_type": str(f.getReturnType()),
            "is_external": f.isExternal(),
            "is_thunk": f.isThunk(),
            "size": body.getNumAddresses() if body is not None else 0,
        })
    write_json(out_dir, "functions.json", funcs)


def dump_imports_exports(program, out_dir):
    symtab = program.getSymbolTable()
    imports = []
    exports = []

    # Imports: external symbols
    ext_it = symtab.getExternalSymbols()
    while ext_it.hasNext():
        sym = ext_it.next()
        imports.append({
            "name": sym.getName(),
            "address": str(sym.getAddress()),
            "source": str(sym.getSource()),
            "library": sym.getLibraryName() if hasattr(sym, "getLibraryName") else None,
        })

    # Exports: primary, non-external functions
    sym_it = symtab.getAllSymbols(True)
    while sym_it.hasNext():
        sym = sym_it.next()
        if sym.getSymbolType() != SymbolType.FUNCTION:
            continue
        if sym.isExternal():
            continue
        if not sym.isPrimary():
            continue
        exports.append({
            "name": sym.getName(),
            "address": str(sym.getAddress()),
            "source": str(sym.getSource()),
        })

    write_json(out_dir, "imports.json", imports)
    write_json(out_dir, "exports.json", exports)


def dump_decompilation(program, out_dir):
    fm = program.getFunctionManager()
    funcs = []
    it = fm.getFunctions(True)
    while it.hasNext():
        f = it.next()
        if f.isExternal():
            continue
        body = f.getBody()
        size = body.getNumAddresses() if body is not None else 0
        funcs.append((size, f))

    funcs.sort(key=lambda x: x[0], reverse=True)
    top_n = 30
    target = [f for (_, f) in funcs[:top_n]]

    if len(target) == 0:
        write_json(out_dir, "decompilation.json", [])
        return

    iface = DecompInterface()
    iface.openProgram(program)
    iface.setTimeout(30)

    out = []
    for f in target:
        res = iface.decompileFunction(f, 30, monitor)
        if res is None or not res.decompileCompleted():
            code = None
        else:
            code = res.getDecompiledFunction().getC()
        out.append({
            "name": f.getName(),
            "entry": str(f.getEntryPoint()),
            "size": f.getBody().getNumAddresses() if f.getBody() is not None else 0,
            "c": code,
        })

    write_json(out_dir, "decompilation.json", out)


args = getScriptArgs()
if args is None or len(args) < 1:
    printerr("At3Dump.py requires output_dir argument")
else:
    out_dir = args[0]
    if not os.path.isabs(out_dir):
        out_dir = os.path.abspath(out_dir)
    if not os.path.isdir(out_dir):
        os.makedirs(out_dir)

    if currentProgram is None:
        printerr("No program loaded")
    else:
        dump_overview(currentProgram, out_dir)
        dump_strings(currentProgram, out_dir)
        dump_functions(currentProgram, out_dir)
        dump_imports_exports(currentProgram, out_dir)
        dump_decompilation(currentProgram, out_dir)

File: reverse/ghidra_scripts/At3Dump.java
// Ghidra Java script: dump analysis artifacts to JSON for AI ingestion
// Usage (headless): -postScript At3Dump.java <output_dir>
//@category Export

import java.io.File;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.data.StringDataInstance;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.symbol.ExternalLocation;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolIterator;
import ghidra.program.model.symbol.SymbolTable;
import ghidra.program.model.symbol.SymbolType;

public class At3Dump extends GhidraScript {

	private Gson gson = new GsonBuilder().setPrettyPrinting().create();

	@Override
	public void run() throws Exception {
		String[] args = getScriptArgs();
		if (args == null || args.length < 1) {
			printerr("At3Dump.java requires output_dir argument");
			return;
		}

		File outDir = new File(args[0]);
		if (!outDir.isAbsolute()) {
			outDir = outDir.getAbsoluteFile();
		}
		if (!outDir.exists()) {
			outDir.mkdirs();
		}

		if (currentProgram == null) {
			printerr("No program loaded");
			return;
		}

		dumpOverview(outDir);
		dumpStrings(outDir);
		dumpFunctions(outDir);
		dumpImportsExports(outDir);
		dumpDecompilation(outDir);
		dumpData(outDir);
	}

	private void dumpData(File outDir) throws Exception {
		Listing listing = currentProgram.getListing();
		DataIterator it = listing.getDefinedData(true);
		List<Map<String, Object>> dataList = new ArrayList<>();
		int maxData = 10000;
		while (it.hasNext() && dataList.size() < maxData && !monitor.isCancelled()) {
			Data data = it.next();
			if (data.isPointer() || data.isString() || data.isStructure()) {
				continue;
			}
			
			String type = data.getDataType().getName().toLowerCase();
			if (type.contains("float") || type.contains("double") || type.contains("undefined4")) {
				Map<String, Object> entry = new HashMap<>();
				entry.put("address", data.getAddress().toString());
				entry.put("type", type);
				entry.put("value", data.getValue() != null ? data.getValue().toString() : "null");
				
				// If it's undefined4, try to interpret as float
				if (type.contains("undefined4")) {
					try {
						int raw = (int)data.getScalar(0).getUnsignedValue();
						float f = Float.intBitsToFloat(raw);
						entry.put("as_float", f);
					} catch (Exception e) {}
				}
				
				dataList.add(entry);
			}
		}
		writeJson(new File(outDir, "data.json"), dataList);
	}

	private void dumpOverview(File outDir) throws Exception {
		Map<String, Object> info = new HashMap<>();
		info.put("name", currentProgram.getName());
		info.put("language", currentProgram.getLanguageID().toString());
		info.put("compiler", currentProgram.getCompilerSpec().getCompilerSpecID().toString());
		info.put("image_base", currentProgram.getImageBase().toString());
		writeJson(new File(outDir, "overview.json"), info);
	}

	private void dumpStrings(File outDir) throws Exception {
		Listing listing = currentProgram.getListing();
		DataIterator it = listing.getDefinedData(true);
		List<Map<String, Object>> strings = new ArrayList<>();
		int maxStrings = 5000;
		while (it.hasNext() && strings.size() < maxStrings && !monitor.isCancelled()) {
			Data data = it.next();
			try {
				StringDataInstance sdi = StringDataInstance.getStringDataInstance(data);
				if (sdi == null) {
					continue;
				}
				String value = sdi.getStringValue();
				if (value == null) {
					continue;
				}
				Map<String, Object> entry = new HashMap<>();
				entry.put("address", data.getAddress().toString());
				entry.put("value", value);
				entry.put("length", sdi.getStringLength());
				strings.add(entry);
			}
			catch (Exception e) {
				// Not a string
			}
		}
		writeJson(new File(outDir, "strings.json"), strings);
	}

	private void dumpFunctions(File outDir) throws Exception {
		FunctionManager fm = currentProgram.getFunctionManager();
		FunctionIterator it = fm.getFunctions(true);
		List<Map<String, Object>> funcs = new ArrayList<>();
		while (it.hasNext() && !monitor.isCancelled()) {
			Function f = it.next();
			Map<String, Object> entry = new HashMap<>();
			entry.put("name", f.getName());
			entry.put("entry", f.getEntryPoint().toString());
			entry.put("signature", f.getSignature().toString());
			entry.put("param_count", f.getParameterCount());
			entry.put("return_type", f.getReturnType().toString());
			entry.put("is_external", f.isExternal());
			entry.put("is_thunk", f.isThunk());
			int size = f.getBody() != null ? (int) f.getBody().getNumAddresses() : 0;
			entry.put("size", size);
			funcs.add(entry);
		}
		writeJson(new File(outDir, "functions.json"), funcs);
	}

	private void dumpImportsExports(File outDir) throws Exception {
		SymbolTable symtab = currentProgram.getSymbolTable();
		List<Map<String, Object>> imports = new ArrayList<>();
		List<Map<String, Object>> exports = new ArrayList<>();

		SymbolIterator extIt = symtab.getExternalSymbols();
		while (extIt.hasNext() && !monitor.isCancelled()) {
			Symbol sym = extIt.next();
			Map<String, Object> entry = new HashMap<>();
			entry.put("name", sym.getName());
			entry.put("address", sym.getAddress().toString());
			entry.put("source", sym.getSource().toString());
			String library = null;
			Object obj = sym.getObject();
			if (obj instanceof ExternalLocation) {
				library = ((ExternalLocation) obj).getLibraryName();
			}
			entry.put("library", library);
			imports.add(entry);
		}

		SymbolIterator symIt = symtab.getAllSymbols(true);
		while (symIt.hasNext() && !monitor.isCancelled()) {
			Symbol sym = symIt.next();
			if (sym.getSymbolType() != SymbolType.FUNCTION) {
				continue;
			}
			if (sym.isExternal()) {
				continue;
			}
			if (!sym.isPrimary()) {
				continue;
			}
			Map<String, Object> entry = new HashMap<>();
			entry.put("name", sym.getName());
			entry.put("address", sym.getAddress().toString());
			entry.put("source", sym.getSource().toString());
			exports.add(entry);
		}

		writeJson(new File(outDir, "imports.json"), imports);
		writeJson(new File(outDir, "exports.json"), exports);
	}

	private void dumpDecompilation(File outDir) throws Exception {
		FunctionManager fm = currentProgram.getFunctionManager();
		FunctionIterator it = fm.getFunctions(true);
		List<Function> funcs = new ArrayList<>();
		while (it.hasNext() && !monitor.isCancelled()) {
			Function f = it.next();
			if (!f.isExternal()) {
				funcs.add(f);
			}
		}

		Collections.sort(funcs, new Comparator<Function>() {
			@Override
			public int compare(Function a, Function b) {
				int sizeA = a.getBody() != null ? (int) a.getBody().getNumAddresses() : 0;
				int sizeB = b.getBody() != null ? (int) b.getBody().getNumAddresses() : 0;
				return Integer.compare(sizeB, sizeA);
			}
		});

		int topN = Math.min(30, funcs.size());
		List<Map<String, Object>> out = new ArrayList<>();
		if (topN == 0) {
			writeJson(new File(outDir, "decompilation.json"), out);
			return;
		}

		DecompInterface iface = new DecompInterface();
		iface.openProgram(currentProgram);

		for (int i = 0; i < topN && !monitor.isCancelled(); i++) {
			Function f = funcs.get(i);
			DecompileResults res = iface.decompileFunction(f, 30, monitor);
			String code = null;
			if (res != null && res.decompileCompleted()) {
				code = res.getDecompiledFunction().getC();
			}

			Map<String, Object> entry = new HashMap<>();
			entry.put("name", f.getName());
			entry.put("entry", f.getEntryPoint().toString());
			int size = f.getBody() != null ? (int) f.getBody().getNumAddresses() : 0;
			entry.put("size", size);
			entry.put("c", code);
			out.add(entry);
		}

		writeJson(new File(outDir, "decompilation.json"), out);
	}

	private void writeJson(File file, Object obj) throws Exception {
		try (FileWriter writer = new FileWriter(file)) {
			gson.toJson(obj, writer);
		}
	}
}
